File: src/net/dinkla/raytracer/objects/utilities/GridUtilities.java
package net.dinkla.raytracer.objects.utilities;

import net.dinkla.raytracer.math.Normal;
import net.dinkla.raytracer.math.Point3DF;
import net.dinkla.raytracer.math.Vector3DF;
import net.dinkla.raytracer.objects.SmoothTriangle;
import net.dinkla.raytracer.objects.Triangle;

import java.util.List;

import static java.lang.Math.sin;
import static java.lang.Math.cos;
import static java.lang.Math.PI;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 16.05.2010
 * Time: 10:19:32
 * To change this template use File | Settings | File Templates.
 */
public class GridUtilities {

    static public void tessellateFlatSphere(List list, final int horizontalSteps, final int verticalSteps) {

        // define the top triangles which all touch the north pole
        int k = 1;

        for (int j = 0; j <= horizontalSteps - 1; j++) {
            // define vertices

            Point3DF v0 = new Point3DF(0, 1, 0); 						// top (north pole)

            Point3DF v1 = new Point3DF((float) (sin(2.0f * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 			// bottom left
                        (float) (cos(PI * k / verticalSteps)),
                        (float) (cos(2.0f * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)));

            Point3DF v2 = new Point3DF((float)(sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 		// bottom  right
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)));

            Triangle triangle_ptr = new Triangle(v0, v1, v2);
            list.add(triangle_ptr);
        }

        // define the bottom triangles which all touch the south pole
        k = verticalSteps - 1;
        for (int j = 0; j <= horizontalSteps - 1; j++) {
            // define vertices

            Point3DF v0 = new Point3DF((float)(sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 			// top left
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

            Point3DF v1 = new Point3DF(0, -1, 0);																		// bottom (south pole)

            Point3DF v2 = new Point3DF((float)(sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 		// top right
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps))	);

            Triangle triangle_ptr = new Triangle(v0, v1, v2);
            list.add(triangle_ptr);
        }

        //  define the other triangles
        for (k = 1; k <= verticalSteps - 2; k++) {
            for (int j = 0; j <= horizontalSteps - 1; j++) {
                // define the first triangle

                // vertices

                Point3DF v0 = new Point3DF((float)(sin(2.0 * PI * j / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 				// bottom left, use k + 1, j
                            (float)(cos(PI * (k + 1) / verticalSteps)),
                            (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))	);

                Point3DF v1 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 		// bottom  right, use k + 1, j + 1
                            (float)(cos(PI * (k + 1) / verticalSteps)),
                            (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)	));

                Point3DF v2 = new Point3DF((float)(	sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 					// top left, 	use k, j
                            (float)(cos(PI * k / verticalSteps)),
                            (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

                Triangle triangle_ptr1 = new Triangle(v0, v1, v2);
                list.add(triangle_ptr1);


                // define the second triangle

                // vertices

                v0 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 			// top right, use k, j + 1
                                (float)(cos(PI * k / verticalSteps)),
                                (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)) );

                v1 = new Point3DF((float)(	sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 				// top left, 	use k, j
                                (float)(cos(PI * k / verticalSteps)),
                                (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

                v2 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 	// bottom  right, use k + 1, j + 1
                                (float)(cos(PI * (k + 1) / verticalSteps)),
                                (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))	);

                Triangle triangle_ptr2 = new Triangle(v0, v1, v2);
                list.add(triangle_ptr2);
            }
        }
    }

    static public void tessellateSmoothSphere(List list, final int horizontalSteps, final int verticalSteps) {
        // define the top triangles which all touch the north pole
        int k = 1;

        for (int j = 0; j <= horizontalSteps - 1; j++) {
            // define vertices

            Point3DF v0 = new Point3DF(0, 1, 0); 						// top (north pole)

            Point3DF v1 = new Point3DF((float) (sin(2.0f * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 			// bottom left
                        (float) (cos(PI * k / verticalSteps)),
                        (float) (cos(2.0f * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)));

            Point3DF v2 = new Point3DF((float)(sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 		// bottom  right
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)));

            SmoothTriangle triangle = new SmoothTriangle(v0, v1, v2);
            triangle.setN0(new Normal(v0.toVector()));
            triangle.setN1(new Normal(v1.toVector()));
            triangle.setN2(new Normal(v2.toVector()));
            list.add(triangle);
        }

        // define the bottom triangles which all touch the south pole
        k = verticalSteps - 1;
        for (int j = 0; j <= horizontalSteps - 1; j++) {
            // define vertices

            Point3DF v0 = new Point3DF((float)(sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 			// top left
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

            Point3DF v1 = new Point3DF(0, -1, 0);																		// bottom (south pole)

            Point3DF v2 = new Point3DF((float)(sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 		// top right
                        (float)(cos(PI * k / verticalSteps)),
                        (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps))	);

            SmoothTriangle triangle = new SmoothTriangle(v0, v1, v2);
            triangle.setN0(new Normal(v0.toVector()));
            triangle.setN1(new Normal(v1.toVector()));
            triangle.setN2(new Normal(v2.toVector()));
            list.add(triangle);
        }

        //  define the other triangles
        for (k = 1; k <= verticalSteps - 2; k++) {
            for (int j = 0; j <= horizontalSteps - 1; j++) {
                // define the first triangle

                // vertices

                Point3DF v0 = new Point3DF((float)(sin(2.0 * PI * j / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 				// bottom left, use k + 1, j
                            (float)(cos(PI * (k + 1) / verticalSteps)),
                            (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))	);

                Point3DF v1 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 		// bottom  right, use k + 1, j + 1
                            (float)(cos(PI * (k + 1) / verticalSteps)),
                            (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)	));

                Point3DF v2 = new Point3DF((float)(	sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 					// top left, 	use k, j
                            (float)(cos(PI * k / verticalSteps)),
                            (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

                SmoothTriangle triangle = new SmoothTriangle(v0, v1, v2);
                triangle.setN0(new Normal(v0.toVector()));
                triangle.setN1(new Normal(v1.toVector()));
                triangle.setN2(new Normal(v2.toVector()));
                list.add(triangle);


                // define the second triangle

                // vertices

                v0 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)), 			// top right, use k, j + 1
                                (float)(cos(PI * k / verticalSteps)),
                                (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * k / verticalSteps)) );

                v1 = new Point3DF((float)(	sin(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)), 				// top left, 	use k, j
                                (float)(cos(PI * k / verticalSteps)),
                                (float)(cos(2.0 * PI * j / horizontalSteps) * sin(PI * k / verticalSteps)	));

                v2 = new Point3DF((float)(	sin(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps)), 	// bottom  right, use k + 1, j + 1
                                (float)(cos(PI * (k + 1) / verticalSteps)),
                                (float)(cos(2.0 * PI * (j + 1) / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))	);

                SmoothTriangle triangle2 = new SmoothTriangle(v0, v1, v2);
                triangle.setN0(new Normal(v0.toVector()));
                triangle.setN1(new Normal(v1.toVector()));
                triangle.setN2(new Normal(v2.toVector()));
                list.add(triangle2);
            }
        }
    }

}

File: src/net/dinkla/raytracer/objects/utilities/GridUtilities.kt
package net.dinkla.raytracer.objects.utilities

import net.dinkla.raytracer.math.Normal
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.objects.SmoothTriangle
import net.dinkla.raytracer.objects.Triangle

import java.lang.Math.sin
import java.lang.Math.cos
import java.lang.Math.PI

object GridUtilities {

    fun tessellateFlatSphere(list: MutableList<Triangle>, horizontalSteps: Int, verticalSteps: Int) {

        // define the top triangles which all touch the north pole
        var k = 1

        for (j in 0..horizontalSteps - 1) {
            // define vertices

            val v0 = Point3D(0.0, 1.0, 0.0)                        // top (north pole)

            val v1 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // bottom left
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // bottom  right
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val triangle_ptr = Triangle(v0, v1, v2)
            list.add(triangle_ptr)
        }

        // define the bottom triangles which all touch the south pole
        k = verticalSteps - 1
        for (j in 0..horizontalSteps - 1) {
            // define vertices

            val v0 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val v1 = Point3D(0.0, -1.0, 0.0)                                                                        // bottom (south pole)

            val v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top right
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val triangle_ptr = Triangle(v0, v1, v2)
            list.add(triangle_ptr)
        }

        //  define the other triangles
        k = 1
        while (k <= verticalSteps - 2) {
            for (j in 0..horizontalSteps - 1) {
                // define the first triangle

                // vertices

                var v0 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom left, use k + 1, j
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                var v1 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom  right, use k + 1, j + 1
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                var v2 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left, 	use k, j
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                val triangle_ptr1 = Triangle(v0, v1, v2)
                list.add(triangle_ptr1)


                // define the second triangle

                // vertices

                v0 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top right, use k, j + 1
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                v1 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left, 	use k, j
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom  right, use k + 1, j + 1
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                val triangle_ptr2 = Triangle(v0, v1, v2)
                list.add(triangle_ptr2)
            }
            k++
        }
    }

    fun tessellateSmoothSphere(list: MutableList<SmoothTriangle>, horizontalSteps: Int, verticalSteps: Int) {
        // define the top triangles which all touch the north pole
        var k = 1

        for (j in 0..horizontalSteps - 1) {
            // define vertices

            val v0 = Point3D(0.0, 1.0, 0.0)                        // top (north pole)

            val v1 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // bottom left
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // bottom  right
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val triangle = SmoothTriangle(v0, v1, v2)
            triangle.n0 = Normal(Vector3D(v0))
            triangle.n1 = Normal(Vector3D(v1))
            triangle.n2 = Normal(Vector3D(v2))
            list.add(triangle)
        }

        // define the bottom triangles which all touch the south pole
        k = verticalSteps - 1
        for (j in 0..horizontalSteps - 1) {
            // define vertices

            val v0 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val v1 = Point3D(0.0, -1.0, 0.0)                                                                        // bottom (south pole)

            val v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top right
                    cos(PI * k / verticalSteps),
                    cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

            val triangle = SmoothTriangle(v0, v1, v2)
            triangle.n0 = Normal(Vector3D(v0))
            triangle.n1 = Normal(Vector3D(v1))
            triangle.n2 = Normal(Vector3D(v2))
            list.add(triangle)
        }

        //  define the other triangles
        k = 1
        while (k <= verticalSteps - 2) {
            for (j in 0..horizontalSteps - 1) {
                // define the first triangle

                // vertices

                var v0 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom left, use k + 1, j
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                var v1 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom  right, use k + 1, j + 1
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                var v2 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left, 	use k, j
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                val triangle = SmoothTriangle(v0, v1, v2)
                triangle.n0 = Normal(Vector3D(v0))
                triangle.n1 = Normal(Vector3D(v1))
                triangle.n2 = Normal(Vector3D(v2))
                list.add(triangle)


                // define the second triangle

                // vertices

                v0 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top right, use k, j + 1
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                v1 = Point3D(sin(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps), // top left, 	use k, j
                        cos(PI * k / verticalSteps),
                        cos(2.0 * PI * j.toDouble() / horizontalSteps) * sin(PI * k / verticalSteps))

                v2 = Point3D(sin(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps), // bottom  right, use k + 1, j + 1
                        cos(PI * (k + 1) / verticalSteps),
                        cos(2.0 * PI * (j + 1).toDouble() / horizontalSteps) * sin(PI * (k + 1) / verticalSteps))

                val triangle2 = SmoothTriangle(v0, v1, v2)
                triangle2.n0 = Normal(Vector3D(v0))
                triangle2.n1 = Normal(Vector3D(v1))
                triangle2.n2 = Normal(Vector3D(v2))
                list.add(triangle2)
            }
            k++
        }
    }

}
